### PR TITLE
Better Rubocop exludes patterns

### PR DESCRIPTION
**Pull request by bf4**
Ref: https://github.com/rubygems/rubygems.org/pull/1019#discussion_r34818587

@qrush 

> I'm not sure if it's worth it to apply rubocop rules to the db directory.

Note, I also excluded `test/` since there were some style issues in there.

### CODE REVIEW

**Comment by qrush** [comment]
I'm fine with checking `test/` - that's easy to ignore but causes problems especially as test frameworks, runners change...which in Ruby is a constant thing. Any feelings on this @sferik ?

**Comment by bf4** [comment]
I mostly chose to skip it because I didn't want there to be failures in the run for this PR.  RSpec, for example, doesn't disable check on spec, but in CI it only runs checks on lib. So, there's that pattern..

**Comment by bf4** [comment]
I pushed up second commit which basically allows for optionally checking tests

**Comment by bf4** [comment]
I don't think it's worth a code comment, but the reason for  `Rake::Task[:rubocop].clear if Rake::Task.task_defined?(:rubocop)` is so that the Rake task we define doesn't enhance the existing default task.

**Comment by bf4** [comment]
@sferik Do we want rubocop to fail the build?  i.e. do we want

```ruby
if ENV["FULL_BUILD"] == "true" # we're on Travis
  # don"t abort rake on failure
  task.fail_on_error = false
end
```

**Comment by bf4** [comment]
I'm not sure what's causing that failure. It doesn't seem important, in any case.

File: .rubocop.yml [hunk 1]
```diff
@@ -2,8 +2,8 @@ inherit_from: .rubocop_todo.yml
 
 AllCops:
   Exclude:
-    - db/schema.rb
-    - vendor/**/*
+    - !ruby/regexp /(Gemfile|Rakefile)$/
```
Comment (sferik) [inline .rubocop.yml]: :-1: Why exclude `Gemfile` and `Rakefile`?

File: .rubocop.yml [hunk 2]
```diff
@@ -2,8 +2,8 @@ inherit_from: .rubocop_todo.yml
...
+    - !ruby/regexp /(vendor|bundle|bin|db|tasks)\/.*/
```
Comment (sferik) [inline .rubocop.yml]: Also, why exclude rake tasks?

**Comment by sferik** [comment]
Let’s exclude files/directories as actual issues arise. I’m :+1: with excluding the `db` directory, since that was an actually issue in https://github.com/rubygems/rubygems.org/pull/1019.

On the other hand, I see no reason to exclude `Gemfile`, `Rakefile`, or the `tasks` directory.

**Comment by sferik** [comment]
> @sferik Do we want rubocop to fail the build?

Yes, I believe we do. The whole point of RuboCop is that RuboCop says “please use Ruby 1.9-style hash syntax instead of rockets” immediately after you open a pull request, so we don’t need to say that manually. RuboCop also has the ability to automatically correct such problems. I’ve added a [section to `CONTRIBUTING.md`](https://github.com/rubygems/rubygems.org/blob/master/CONTRIBUTING.md#running-rubocop) about this, so people will hopefully catch these issues even *before* they open a pull request but failing the build is a pretty good backstop.

**Comment by bf4** [comment]
> 
Awesome.  I'll make those changes presently.  Do we want to check tests on ci? They fail now is all
>

**Comment by sferik** [comment]
> Do we want to check tests on ci? They fail now is all

Try rebasing from `master`. It should be :green_heart: now.

**Comment by bf4** [comment]
Also, one of my 'failing' rubocop tests was an untracked file from a different pr hanging around (oops)

**Comment by bf4** [comment]
If the diff looks good, I'll rebase

File: lib/tasks/rubocop.rake [hunk 3]
```diff
@@ -1,8 +1,21 @@
...
+    task.formatters = ['progress']
+    task.options = ['--rails', '--display-cop-names', '--display-style-guide']
+    task.patterns = [
+      'Gemfile',
+      'Rakefile',
+      'lib/**/*.{rb,rake}',
+      'config/**/*.rb',
+      'app/**/*.rb',
+      'test/**/*.rb'
+    ]
```
Comment (bf4) [inline lib/tasks/rubocop.rake]: these patterns overrides whatever's in the `.rubocop.yml`. I've found it convenient to do it this way in the past, but don't feel strongly about it if you'd prefer to only have one definition. As I wrote in one of the commits, we could have less strict Rubocop behavior on CI, e.g. we could have

> bundle exec rake rubocop # won't run on test/
> bundle exec rubocop         # will run on test/

**Comment by sferik** [comment]
Final request (from me): would you mind squashing your three commits into a single one with a clear commit message?

**Comment by bf4** [comment]
@sferik of course.  I just added the new one to make the changes easier to see for the reviewer. :) I'm big on the rebase workflow :)

**Comment by sferik** [comment]
Thanks Benjamin!

**Pull request merged**